Sources/OpenAPIRuntime/Multipart/MultipartParser.swift

```diff
@@ -0,0 +1,342 @@
+                state = .mutating
+                switch partState {
+                case .parsingHeaderFields(var headerFields):
+                    // Either we find `--` in which case there are no more parts and we're finished, or something else
+                    // and we start parsing headers.
+                    switch buffer.firstIndexAfterElements(ASCII.dashes) {
+                    case .index(let index):
+                        state = .finished
+                        buffer.removeSubrange(..<index)
+                        return .returnNil
+                    case .reachedEndOfSelf:
+                        state = .parsingPart(buffer, .parsingHeaderFields(headerFields))
+                        return .needsMore
+                    case .mismatchedCharacter: break
+                    }
+                    // Consume CRLF
+                    let indexAfterFirstCRLF: [UInt8].Index
+                    switch buffer.firstIndexAfterElements(ASCII.crlf) {
+                    case .index(let index): indexAfterFirstCRLF = index
+                    case .reachedEndOfSelf:
+                        state = .parsingPart(buffer, .parsingHeaderFields(headerFields))
+                        return .needsMore
+                    case .mismatchedCharacter:
+                        state = .finished
+                        return .emitError(.invalidCRLFAtStartOfHeaderField)
+                    }
+                    // If CRLF is here, this is the end of header fields section.
+                    switch buffer[indexAfterFirstCRLF...].firstIndexAfterElements(ASCII.crlf) {
+                    case .index(let index):
+                        buffer.removeSubrange(buffer.startIndex..<index)
+                        state = .parsingPart(buffer, .parsingBody)
+                        return .emitHeaderFields(headerFields)
+                    case .reachedEndOfSelf:
+                        state = .parsingPart(buffer, .parsingHeaderFields(headerFields))
+                        return .needsMore
+                    case .mismatchedCharacter: break
+                    }
+                    let startHeaderNameIndex = indexAfterFirstCRLF
+                    guard
+                        let endHeaderNameIndex = buffer[startHeaderNameIndex...]
+                            .firstIndex(where: { !ASCII.isValidHeaderFieldNameByte($0) })
+                    else {
+                        // No index matched yet, we need more data.
+                        state = .parsingPart(buffer, .parsingHeaderFields(headerFields))
+                        return .needsMore
+                    }
+                    let startHeaderValueWithWhitespaceIndex: [UInt8].Index
+                    // Check that what follows is a colon, otherwise this is a malformed header field line.
+                    // Source: RFC 7230, section 3.2.4.
+                    switch buffer[endHeaderNameIndex...].firstIndexAfterElements([ASCII.colon]) {
```

Comment:
~I assume this is a check for `No whitespace is allowed between the header field-name and colon.`, however, this just finds the first index after the next colon and if it finds one, then it sets that as the start of the header value. FWICT it only returns `.emitError(.missingColonAfterHeaderName)` if there is no colon.~

I wrote all of that with a false assumption of what `firstIndexAfterElements` does, which, I think could do with renaming. From the name alone, I assumed it would return the first index after the elements if they exist anywhere in the remaining sequence, not just at the start.

Reply:
Good point, ideas for a better name?

Reply:
It looks like this is usually followed by removing these elements from the front of the buffer. Do you think we could express this in some form of the term "pop", e.g. `popPrefix`. Maybe it would be mutating and return an enum value, simliar to how you do now. If the prefix _is_ there it is returned (and dropped from the collection), if it isn't you get either `.reachedEnd` or `.unexpectedPrefix`.

Then you can progress through the bytes that you have with e.g. `buffer.popPrefix([colon])`

Reply:
There are cases where we keep reading more bytes before dropping the prefix, so I wouldn't turn it into a mutating function, but I like your suggestion around the "prefix" terminology - that's much better. Let me update it.

Reply:
Updated: https://github.com/apple/swift-openapi-runtime/pull/72/commits/115734250be5359adc7e617e471e28f6a3cf1de8